src/renderers/dom/client/wrappers/ReactDOMInput.js

```diff
@@ -151,6 +157,41 @@ var ReactDOMInput = {
 
     if (__DEV__) {
       warnIfValueIsNull(props);
+
+      var initialValue = inst._wrapperState.initialChecked || inst._wrapperState.initialValue;
+      var defaultValue = props.defaultChecked || props.defaultValue;
+      var controlled = props.checked !== undefined || props.value !== undefined;
+      var owner = inst._currentElement._owner;
+
+      if (
+        (initialValue || !inst._wrapperState.controlled) &&
+        controlled && !didWarnUncontrolledToControlled
+      ) {
+        warning(
+          false,
+          '%s is changing a uncontrolled input to be controlled. ' +
+          'Input elements should not switch from uncontrolled to controlled(or viceversa). ' +
+          'Decide between using a controlled or uncontrolled input ' +
+          'element for the lifetime of the component. More info: https://fb.me/react-controlled-components',
+          owner && owner.getName() || 'A component'
+        );
+        didWarnUncontrolledToControlled = true;
+      }
+      if (
+        inst._wrapperState.controlled &&
+        (defaultValue || !controlled) &&
+        !didWarnControlledToUncontrolled
+      ) {
+        warning(
+          false,
+          '%s is changing a controlled input to be uncontrolled. ' +
```

Comment:
Can we also specify the type of the input (checkbox/radio/etc).  We can just print the `props.type` to do that.